src/lib/pubkey/curve448/curve448_utils/curve448_gf.cpp

```diff
@@ -0,0 +1,399 @@
+ */
+void gf_sub(std::span<uint64_t, 7> out, std::span<const uint64_t, 7> a, std::span<const uint64_t, 7> b) {
+   std::array<uint64_t, 7> h_0;
+   std::array<uint64_t, 7> h_1;
+
+   bool borrow = false;
+   for(size_t i = 0; i < 7; ++i) {
+      h_0[i] = u64_sub_with_borrow(a[i], b[i], &borrow);
+   }
+   uint64_t delta = borrow;
+   uint64_t delta_p = delta << 32;
+   borrow = false;
+
+   h_1[0] = u64_sub_with_borrow(h_0[0], delta, &borrow);
+   h_1[1] = u64_sub_with_borrow(h_0[1], 0, &borrow);
+   h_1[2] = u64_sub_with_borrow(h_0[2], 0, &borrow);
+   h_1[3] = u64_sub_with_borrow(h_0[3], delta_p, &borrow);
+   h_1[4] = u64_sub_with_borrow(h_0[4], 0, &borrow);
+   h_1[5] = u64_sub_with_borrow(h_0[5], 0, &borrow);
+   h_1[6] = u64_sub_with_borrow(h_0[6], 0, &borrow);
+
+   delta = borrow;
+   delta_p = delta << 32;
+   borrow = false;
+
+   out[0] = u64_sub_with_borrow(h_1[0], delta, &borrow);
+   out[1] = u64_sub_with_borrow(h_1[1], 0, &borrow);
+   out[2] = u64_sub_with_borrow(h_1[2], 0, &borrow);
+   out[3] = u64_sub_with_borrow(h_1[3], delta_p, &borrow);
+   out[4] = h_1[4];
+   out[5] = h_1[5];
+   out[6] = h_1[6];
+}
+
+/**
+ * @brief Inversion in GF(P) using Fermat's little theorem:
+ * x^-1 = x^(P-2) mod P
+ */
+void gf_inv(std::span<uint64_t, 7> out, std::span<const uint64_t, 7> a) {
+   clear_mem(out);
+   out[0] = 1;
+   // Square and multiply
+   for(int16_t t = 448; t >= 0; --t) {
+      gf_square(out, out);
+      // (P-2) has zero bits at indices 1, 224, 448. All others are one.
+      if(t != 448 && t != 224 && t != 1) {
+         gf_mul(out, out, a);
+      }
+   }
+}
+
+/**
+ * @brief Convert a number to its canonical representation.
+ *
+ * I.e. if the number is greater than P, subtract P. The number cannot be >= 2P
+ * since 2*P > 2^(7*64).
+ */
+std::array<uint64_t, 7> to_canonical(std::span<const uint64_t, 7> in) {
+   std::array<uint64_t, 7> p = {0xffffffffffffffff,
+                                0xffffffffffffffff,
+                                0xffffffffffffffff,
+                                0xfffffffeffffffff,
+                                0xffffffffffffffff,
+                                0xffffffffffffffff,
+                                0xffffffffffffffff};
+
+   std::array<uint64_t, 7> in_minus_p;
+   bool borrow = false;
+   for(int i = 0; i < 7; ++i) {
+      in_minus_p[i] = u64_sub_with_borrow(in[i], p[i], &borrow);
+   }
+   std::array<uint64_t, 7> out;
+   CT::Mask<uint64_t>::expand(borrow).select_n(out.data(), in.data(), in_minus_p.data(), 7);
+   return out;
+}
+
+}  // namespace
+
+Gf448Elem::Gf448Elem(std::span<const uint8_t, 56> x) {
+   load_le(m_x.data(), x.data(), m_x.size());
+}
+
+Gf448Elem::Gf448Elem(uint64_t least_sig_word) {
+   clear_mem(m_x);
+   m_x[0] = least_sig_word;
+}
+
+void Gf448Elem::to_bytes(std::span<uint8_t, 56> out) const {
+   const auto canonical_form = to_canonical(m_x);
+   copy_out_le(out.data(), out.size(), canonical_form.data());
```

Comment:
```suggestion
   store_le(out, to_canonical(m_x);
```

`copy_out_le()` is needed if the buffer lengths don't match (it truncates if not enough space is available). But here we know statically that they do